ui/StatusQ/src/leftjoinmodel.cpp

```diff
@@ -0,0 +1,219 @@
+            || rightModelJoinRoleList.size() != 1) {
+        qWarning().noquote() << QString("Both left and right models have to "
+                                        "contain join role %1!").arg(m_joinRole);
+        return;
+    }
+
+    m_leftModelJoinRole = leftModelJoinRoleList.at(0);
+    m_rightModelJoinRole = rightModelJoinRoleList.at(0);
+
+    auto leftRoles = leftRoleNames.keys();
+    auto maxLeftRole = std::max_element(leftRoles.cbegin(), leftRoles.cend());
+    auto rightRolesOffset = *maxLeftRole + 1;
+    auto roleNames = leftRoleNames;
+
+    auto i = rightRoleNames.constBegin();
+    while (i != rightRoleNames.constEnd()) {
+        if (i.value() != m_joinRole) {
+            auto roleWithOffset = i.key() + rightRolesOffset;
+            roleNames.insert(roleWithOffset, i.value());
+            m_joinedRoles.append(roleWithOffset);
+        }
+        ++i;
+    }
+
+    m_rightModelRolesOffset = rightRolesOffset;
+    m_roleNames = roleNames;
+
+    connect(m_rightModel, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles) {
+        QVector<int> rolesTranslated;
+
+        if (roles.contains(m_rightModelJoinRole)) {
+            rolesTranslated = m_joinedRoles;
+        } else {
+            rolesTranslated = roles;
+
+            for (auto& role : rolesTranslated)
+                role += m_rightModelRolesOffset;
+        }
+
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), rolesTranslated);
+    });
+
+    disconnect(m_leftModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+    disconnect(m_rightModel, &QAbstractItemModel::rowsInserted,
+               this, &LeftJoinModel::initializeIfReady);
+
+    connect(m_rightModel, &QAbstractItemModel::rowsRemoved, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(m_rightModel, &QAbstractItemModel::rowsInserted, this, [this] {
+        emit dataChanged(index(0, 0), index(rowCount() - 1, 0), m_joinedRoles);
+    });
+
+    connect(this, &QAbstractItemModel::dataChanged, this,
+            [this](auto& topLeft, auto& bottomRight, auto& roles ) {
+        if (roles.contains(m_leftModelJoinRole))
+            emit dataChanged(topLeft, bottomRight, m_joinedRoles);
+    });
+
+    QIdentityProxyModel::setSourceModel(m_leftModel);
+}
+
+QVariant LeftJoinModel::data(const QModelIndex& index, int role) const
+{
+    auto idx = m_leftModel->index(index.row(), 0);
+
+    if (role < m_rightModelRolesOffset)
+        return m_leftModel->data(idx, role);
+
+    QVariant joinRoleLeftValue = m_leftModel->data(idx, m_leftModelJoinRole);
+
+    if (m_lastUsedRightModelIndex.isValid()
+            && m_rightModel->data(m_lastUsedRightModelIndex,
+                                  m_rightModelJoinRole) == joinRoleLeftValue)
+    {
+        return m_rightModel->data(m_lastUsedRightModelIndex,
```

Comment:
Impressed of how well this is working. I see 2x more cached index returns than searches in the right model.

Reply:
Do you have an explanation why this is happening? I'm probably missing something because I didn't expect this 😄

Reply:
Because to visualize given index, ListView (or other view) calls `data(index, role)` multiple times using the same index but different roles - all that are referred in a delegate. So for first role we do full search, for next roles we take position from cache.